staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_proxy.go

```diff
@@ -248,37 +244,38 @@ func (r *proxyHandler) updateAPIService(apiService *apiregistrationv1api.APIServ
 		return
 	}
 
-	newInfo := proxyHandlingInfo{
-		name: apiService.Name,
-		restConfig: &restclient.Config{
-			TLSClientConfig: restclient.TLSClientConfig{
-				Insecure:   apiService.Spec.InsecureSkipTLSVerify,
-				ServerName: apiService.Spec.Service.Name + "." + apiService.Spec.Service.Namespace + ".svc",
-				CertData:   r.proxyClientCert,
-				KeyData:    r.proxyClientKey,
-				CAData:     apiService.Spec.CABundle,
+	transportConfig := &transport.Config{
+		TLS: transport.TLSConfig{
+			Insecure: apiService.Spec.InsecureSkipTLSVerify,
+			GetCert: func() (*tls.Certificate, error) {
```

Comment:
every time we create a GetCert function, we create a unique config that busts the cache of reusable client configs (the cache key for this is computed based on the function pointer address, see https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/transport/cache.go#L134)

in contrast, the previous version set static options that reused the cached tls config (and the existing kept-alive backend connections)

The TLS config cache has no size limit, which means this is effectively introducing a slow memory leak if APIService conditions flutter.

Reply:
Hmm does the pointer address change though? 🤔This is implemented with the idea/impression that it does not change, am I wrong?

To test this, I wrote a following test:
- created a setup similar to the TestProxyCertReload
- checked the address in cache keey after calling updateAPIService
- wrote new certs to disk
- called updateAPIService again
- checked the address in cache

The address was the same.

Reply:
@liggitt I've opened https://github.com/kubernetes/kubernetes/pull/92765/commits/adcd6e10406e4c2b791f088d21d249d4350c4e9b that demonstrates the key returned from the cache is the same as long as the address of `GetCert` function doesn't change

Reply:
we could extend `certKeyContentProvider` to provide `GetCert` function so that the address stays the same and change the cache to take into account the content, not the address.

Reply:
I've opened https://github.com/kubernetes/kubernetes/pull/92765/commits/8520587e9c37be6edded2b2ac9f2555614f4110b to demonstrate that. It still creates a new transport for each request. Most of the time the transport will be retrieved from the cache but even that incurs a small performance cost of acquiring a lock.

We could move the code back to `updateAPIService` method and reload the handler on the certificate change. In that case, this approach seems to be more complicated than https://github.com/kubernetes/kubernetes/pull/92791.

Reply:
> @liggitt I've opened [adcd6e1](https://github.com/kubernetes/kubernetes/commit/adcd6e10406e4c2b791f088d21d249d4350c4e9b) that demonstrates the key returned from the cache is the same as long as the address of `GetCert` function doesn't change

hmm... if a function with the same pointer address can return different data to different callers, that means some of the existing ways we build the client config cache could have problems.

> @liggitt we could extend certKeyContentProvider to provide GetCert function so that the address stays the same and change the cache to take into account the content, not the address.

Wouldn't that compute the cache key based on the initial cert content, which would conflate client configurations that should be kept distinct? For example, config A and config B could point to different file paths `/a.ca` and `/b.ca` which happen to contain the same content at startup... if those are being reloaded dynamically, we wouldn't want to make those share a client, since at a later point in time, those files' content could diverge.

Reply:
> hmm... if a function with the same pointer address can return different data to different callers, that means some of the existing ways we build the client config cache could have problems.

comparing the address of a function doesn't seem right since golang forbids comparing the addresses of functions in general.

would you accept [a patch](https://github.com/kubernetes/kubernetes/pull/92765/commits/8520587e9c37be6edded2b2ac9f2555614f4110b#diff-c02000c0f8c4dbc287f36f9e2763d0f3) that changes this behavior? Ideally, if we could serialize `getCert` and compare the bytes.

Reply:
> comparing the address of a function doesn't seem right since golang forbids comparing the addresses of functions in general.

yeah, I'm not thrilled with the approach... the places we do that currently are for non-serializable injected implementations.

> would you accept a patch that changes this behavior? Ideally, if we could serialize getCert and compare the bytes.

we'd need a way to determine the identity of the dynamic methods (GetCert, Dial, and Proxy)